src/aes128gcm.rs

```diff
@@ -23,174 +28,386 @@ const ECE_WEBPUSH_IKM_LENGTH: usize = 32;
-        true
+    /// Get the size occupied by this header when written to the encrypted data.
+    ///
+    pub fn encoded_size(&self) -> usize {
+        ECE_AES128GCM_HEADER_LENGTH + self.keyid.len()
     }
+}
+
+/// Struct representing an individual plaintext record.
+///
+/// The encryption process splits up the input plaintext to fixed-size records,
+/// each of which is encrypted independently. This struct encapsulates all the
+/// data about a particular record. This diagram from the RFC may help you to
+/// visualize how this data gets encrypted:
+///
+///   +-----------+             content
+///   |   data    |             any length up to rs-17 octets
+///   +-----------+
+///        |
+///        v
+///   +-----------+-----+       add a delimiter octet (0x01 or 0x02)
+///   |   data    | pad |       then 0x00-valued octets to rs-16
+///   +-----------+-----+       (or less on the last record)
+///            |
+///            v
+///   +--------------------+    encrypt with AEAD_AES_128_GCM;
+///   |    ciphertext      |    final size is rs;
+///   +--------------------+    the last record can be smaller
+///
+/// To avoid copying data when chunking a plaintext into multiple records, this struct
+/// stores a reference to its portion of the plaintext, borrowed from the underlying data.
+///
+struct PlaintextRecord<'a> {
+    /// The plaintext, to go at the start of the record.
+    plaintext: &'a [u8],
+    /// The amount of padding to be added to the end of the record.
+    /// Always >= 1 in practice, because the first byte of padding is a delimiter.
+    padding: usize,
+    /// The position of this record in the overall sequence of records for some data.
+    sequence_number: usize,
+    /// Whether this is the final record in the data.
+    is_final: bool,
+}
 
-    fn pad_size() -> usize {
-        ECE_AES128GCM_PAD_SIZE
+impl<'a> PlaintextRecord<'a> {
+    /// Decrypt a single record from the given ciphertext, into its corresponding plaintext.
+    ///
+    /// The caller must provide a buffer with sufficient space to store the decrypted plaintext,
+    /// and this method will panic (via Rust's runtime safety checks) if there is insufficient
+    /// space available.
+    ///
+    pub(crate) fn decrypt_from(
+        cryptographer: &dyn Cryptographer,
+        key: &[u8],
+        nonce: &[u8],
+        sequence_number: usize,
+        ciphertext: &[u8],
+        plaintext_buffer: &'a mut [u8],
+    ) -> Result<PlaintextRecord<'a>> {
+        if ciphertext.len() <= ECE_TAG_LENGTH {
+            return Err(Error::BlockTooShort);
+        }
+        let iv = generate_iv_for_record(&nonce, sequence_number);
+        let padded_plaintext = cryptographer.aes_gcm_128_decrypt(&key, &iv, &ciphertext)?;
+        // Scan backwards for the first non-zero byte from the end of the data, which delimits the padding.
+        let last_nonzero_byte = match padded_plaintext.iter().rposition(|&b| b != 0u8) {
```

Comment:
clippy also prefers `if let Some(pos) = ` here and I've learned to prefer it.  But it recently started going with this instead:

```rust
let last_nonzero_idx = padded_plaintext.iter().rposition(|&b| b != 0u8).ok_or(Error::DecryptPadding)?
```
(note that this is an index and not a byte)